BayesianEstimation/AdaptiveMetropolisWithinGibbs.cs

```diff
@@ -0,0 +1,285 @@
+        private double sdMu; // sd of prior prob mean normal distribution
+        private double sigmaLow; // lower bound of prior prob sd uniform distribution
+        private double sigmaHigh; // upper bound of prior prob sd uniform distribution
+        private double exp; // exponent of prior prob exponential distribution
+
+        /// <summary>
+        /// Construct the adaptive Metropolis within Gibbs sampler. Leave data2 null for one-sample data.
+        /// </summary>
+        public AdaptiveMetropolisWithinGibbs(List<double> data1, List<double> data2, int batch_size = 3, int? seed = null)
+        {
+            if (seed != null)
+            {
+                random = new MersenneTwister(seed.Value, true);
+            }
+            else
+            {
+                random = new MersenneTwister(true);
+            }
+
+            batchCount = 0;
+            this.batchSize = batch_size;
+            isTwoSample = data2 != null;
+
+            if (isTwoSample)
+            {
+                Data = new List<double>[] { data1, data2 };
+            }
+            else
+            {
+                Data = new List<double>[] { data1 };
+            }
+            
+            SetUpAcceptableParameterRanges();
+            SetUpPriorProbabilityDistributions();
+            SetUpPosteriorFunction(isTwoSample);
+        }
+
+        /// <summary>
+        /// Burn in and then sample the MCMC chain.
+        /// </summary>
+        public void Run(int burnin = 20000, int n = 20000)
+        {
+            n_samples(burnin);
+            n_samples(n);
+        }
+
+        /// <summary>
+        /// Sets up the acceptable range of values for each parameter. For example, standard deviation must be positive.
+        /// </summary>
+        protected void SetUpAcceptableParameterRanges()
+        {
+            if (acceptableParameterRanges == null)
+            {
+                if (isTwoSample)
+                {
+                    acceptableParameterRanges = new List<Tuple<double, double>>[5];
+                    acceptableParameterRanges[0] = new List<Tuple<double, double>> { new Tuple<double, double>(double.NegativeInfinity, double.PositiveInfinity) }; // mu1 (can take any value)
+                    acceptableParameterRanges[1] = new List<Tuple<double, double>> { new Tuple<double, double>(double.NegativeInfinity, double.PositiveInfinity) }; // mu2 (can take any value)
+                    acceptableParameterRanges[2] = new List<Tuple<double, double>> { new Tuple<double, double>(0, double.PositiveInfinity) }; // sd1 (must be positive)
+                    acceptableParameterRanges[3] = new List<Tuple<double, double>> { new Tuple<double, double>(0, double.PositiveInfinity) }; // sd2 (must be positive)
+                    acceptableParameterRanges[4] = new List<Tuple<double, double>> { new Tuple<double, double>(0, double.PositiveInfinity) }; // nu (must be positive)
```

Comment:
what is "nu"?

Reply:
nu is the degree of normality of the t-distribution. "degrees of freedom" refers to nu.